### PR TITLE
Replace (non-data) TensorUtils calls with non-generic THCTensor calls.

### PR DESCRIPTION
**Pull request by gchanan**
TensorUtils is templatized on the THTensor type, so to support a single tensor type (like ATen), we need to remove these.

This PR does the following:
1) Allows THCTensorTypeUtils.cuh to include THCTensor.hpp.
   This involves moving includes of it outside of generic/, so we can use the new implementations.

2) Defines a single _THTensor struct and changes THCRealTensor to be a derived type of _THCTensor.
   This allows us to implement a single non-generic function and avoid static_cast or void * tricks to call it from the generic functions.

3) For functions inside of TensorUtils that don't use data pointers:
   a) Implement the functions in (non-generic) THTensor.cpp and declare them in (non-generic) THTensor.hpp.
   b) Have the generic versions call the non-generic versions.
   c) Replace the corresponding TensorUtils<THCTensor>::fn call with (non-generic) THTensor_fn.

### CODE REVIEW

**Comment by gchanan** [comment]
Arguable some of these functions should be in different files (e.g. the indexing related ones), but I don't think these functions will stick around long term (once THC is fully ATen-ified they can just call the ATen versions).

File: aten/src/THC/generic/THCTensor.hpp [hunk 1]
```diff
@@ -2,18 +2,7 @@
...
-    int64_t *stride;
-    int nDimension;
-
-    THCStorage *storage;
-    ptrdiff_t storageOffset;
-    std::atomic<int> refcount;
-
-    char flag;
-
+typedef struct THCTensor : _THCTensor {
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensor.hpp]: Would be nice to have a comment here, but since it's dying tomorrow I guess it's fine :)

**Review by ezyang — APPROVED** [review]
Music to my eyes!

**Comment by ezyang** [comment]
```
23:22:04 Running test_nn ...
23:22:06 Traceback (most recent call last):
23:22:06   File "test_nn.py", line 32, in <module>
23:22:06     from common_cuda import TEST_CUDA, TEST_MULTIGPU, TEST_CUDNN, \
23:22:06   File "/var/lib/jenkins/workspace/test/common_cuda.py", line 10, in <module>
23:22:06     TEST_CUDNN = TEST_CUDA and torch.backends.cudnn.is_acceptable(torch.tensor(1., device=CUDA_DEVICE))
23:22:06 RuntimeError: Attempt to access Storage having data type Byte as data type Double
23:22:06 Traceback (most recent call last):
23:22:06   File "test/run_test.py", line 344, in <module>
23:22:06     main()
23:22:06   File "test/run_test.py", line 336, in main
23:22:06     raise RuntimeError(message)
23:22:06 RuntimeError: test_nn failed!
```

I didn't see an obvious mess up in the code, debugging will be necessary, I guess :(

**Comment by gchanan** [comment]
caffe2 failure looks unrelated.

**Pull request merged**